bpf-gpl/xdp.c

```diff
@@ -0,0 +1,93 @@
+#include <linux/if_ether.h>
+#include <linux/ip.h>
+#include <linux/in.h>
+#include <linux/icmp.h>
+#include <linux/tcp.h>
+#include <linux/udp.h>
+
+// stdbool.h has no deps so it's OK to include; stdint.h pulls in parts
+// of the std lib that aren't compatible with BPF.
+#include <stdbool.h>
+
+#include "bpf.h"
+#include "types.h"
+#include "log.h"
+#include "skb.h"
+#include "routes.h"
+#include "reasons.h"
+#include "icmp.h"
+#include "fib.h"
+#include "parsing.h"
+#include "failsafe.h"
+#include "jump.h"
+
+struct flow_tuple {
+	__be32 saddr;
+	__be32 daddr;
+	__be16 sport;
+	__be16 dport;
+	__u8 protocol;
+};
+
+struct bpf_map_def_extended SEC("maps") flowtracker = {
+	.type = BPF_MAP_TYPE_HASH,
+	.key_size = sizeof(struct flow_tuple),
+	.value_size = sizeof(__u64),
+	.max_entries = 32,
+#ifndef __BPFTOOL_LOADER__
+	.pinning_strategy = MAP_PIN_GLOBAL,
+#endif
+};
+
+SEC("prog")
+static CALI_BPF_INLINE int calico_xdp(struct xdp_md *xdp_ctx) {
+
+	struct cali_tc_ctx ctx = {
+		.state = state_get(),
+		.xdp = xdp_ctx,
+		.fwd = {
+			.res = XDP_PASS, // or XDP_DROP?
```

Comment:
Hmm, I think it depends on the wider design.  Given that an XDP program can be installed at several levels, and some hardware may have differing capabilities, what's our plan for ensuring that a program can work at the place where it is installed?

Reply:
I think that should be check in the control plane. For example, if we use "ip" command to load the XDP programs (like TC programs), we can try to load with native support and when fails fall back to the generic mode. 
As you said, the default action depends on the wider design. I'll keep it as "XDP_PASS" since it is safest at the moment, and we should change it later on based on the final design.